polynote-kernel/src/main/scala/polynote/kernel/util/TaskManager.scala

```diff
@@ -1,14 +1,160 @@
+  private val queuedTasks = new LinkedBlockingQueue[TaskManager.Queued[_]]()
+
+  private def publish(taskInfo: TaskInfo): IO[Unit] = statusUpdates.publish1(UpdatedTasks(taskInfo :: Nil))
+
+  private def run[A](taskInfo: TaskInfo, task: IO[A]): IO[Running[A]] = {
+    for {
+      _       <- publish(taskInfo)
+      fiber   <- task.start
+      runTask  = Running(taskInfo, fiber)
+      _       <- IO(running.add(runTask))
+    } yield runTask
+  }
+
+  private def runNext(): IO[Unit] = (IO(queuedTasks.take()) <* semaphore.acquire).flatMap {
+    queued =>
+      System.err.println(s"Running ${queued.taskInfo.label}")
+      run(queued.taskInfo.running, queued.io).flatMap {
+        runTask =>
+          def finalizer: IO[Unit] =
+            IO(running.remove(runTask)) *>
+              publish(queued.taskInfo.completed) *>
+                (runNext().start >>= runQueueFiber.set) *>
+                  semaphore.release
+
+          runTask.fiber.join.uncancelable.flatMap {
+            result =>
+              queued.onComplete(result, finalizer)
+          }.attempt.flatMap {
+            result =>
+              queued.deferred.complete(result) *> {
+                result match {
+                  case Left(err) => finalizer
+                  case Right(_)  => IO.unit
+                }
+              }
+          }
+      }
+  }
+
+  private val runQueueFiber: Ref[IO, Fiber[IO, Unit]] = Ref.unsafe(runNext().start.unsafeRunSync())
+
+  /**
+    * Cancel all queued tasks
+    */
+  def cancelAllQueued: IO[Unit] = IO {
+    val javaList = new java.util.ArrayList[Queued[_]](queuedTasks.size())
+    queuedTasks.drainTo(javaList)
+    javaList.asScala.toList
+  }.flatMap {
+    dequeuedTasks => dequeuedTasks.map {
+      queued => publish(queued.taskInfo.completed)
+    }.sequence.as(())
+  }
+
+  /**
+    * Queue a task which will result in a Stream. The next queued task will not start until the resulting stream
+    * terminates.
+    */
+  def queueTaskStream[A](id: String, label: String, detail: String = "")(fn: TaskInfo => IO[Stream[IO, A]]): IO[IO[Stream[IO, A]]] = {
+    val taskInfo = TaskInfo(id, label, detail, TaskStatus.Queued)
+    publish(taskInfo) *> Deferred[IO, Either[Throwable, Stream[IO, A]]].flatMap {
+      deferred => IO(queuedTasks.put(QueuedStream(taskInfo, fn(taskInfo), deferred))).map {
+        _ => deferred.get.flatMap {
+          case Left(err) => IO.raiseError(err)
+          case Right(result) => IO.pure(result)
+        }
+      }
+    }
+  }
+
+  /**
+    * Run a task by passing a function which will receive the [[TaskInfo]] and return a value. This bypasses the
+    * task queue.
+    *
+    * @return an IO, which will upon evaluation will "block" until the task runs and completes.
+    */
+  def runTask[A](id: String, label: String, detail: String = "")(fn: TaskInfo => A): IO[A] =
+    runTaskIO(id, label, detail)((taskInfo: TaskInfo) => IO.pure(taskInfo).map(fn))
```

Comment:
Just curious about the `map` in `IO.pure(taskInfo).map(fn))`. 
- I guess `IO.pure(fn(taskInfo))` is wrong because `IO.pure` is meant for already computer values (`fn` could have side-effects)
- but why not `IO(fn(taskInfo))`?

Reply:
They're exactly equivalent, actually (from the functor laws)... I kind of like your phrasing better. Will change it.

Reply:
Actually, I think I'm going to leave it as-is. Semantically it's equivalent, but based on the internals I think it's better to use `IO.pure(a).map(fn)` – it creates a `Map` structure instead of a `Delay` structure, which avoids creating an extra thunk (and thus avoids creating an extra classfile for the thunk). Not a big deal, but might as well avoid creating class files where possible.

Reply:
Ah, interesting - I guess we should consider doing that in other parts of the codebase then - I think there are probably quite a few places where we allocate more thunks than needed.